cdap-api/src/main/java/co/cask/cdap/api/plugin/PluginContext.java

```
@@ -60,4 +61,23 @@
    * @throws UnsupportedOperationException if the program does not support plugin
    */
   <T> T newPluginInstance(String pluginId) throws InstantiationException;
+
+  /**
+   * Creates a new instance of a plugin. The instance returned will have the {@link PluginConfig} setup with
+   * {@link PluginProperties} provided at the time when the
+   * {@link PluginConfigurer#usePlugin(String, String, String, PluginProperties)} was called during the
+   * program configuration time. If a plugin field has a macro,
+   * the parameter evaluator is used to evaluate the macro and the evaluated value is used for the plugin field.
+   *
+   * @param pluginId the unique identifier provide when declaring plugin usage in the program.
+   * @param evaluator the macro evaluator that's used to evaluate macro for plugin field
+   *                  if macro is supported on those fields.
+   * @param <T> the class type of the plugin
+   * @return A new instance of the plugin being specified by the arguments
+   *
+   * @throws InstantiationException if failed create a new instance
+   * @throws IllegalArgumentException if pluginId is not found
+   * @throws UnsupportedOperationException if the program does not support plugin
+   */
+  <T> T newPluginInstance(String pluginId, MacroEvaluator evaluator) throws InstantiationException;
```

should this also throw InvalidMacroException?